core/src/main/java/org/elasticsearch/index/store/Store.java

```diff
@@ -657,8 +654,8 @@ final void verifyAfterCleanup(MetadataSnapshot sourceMetaData, MetadataSnapshot
                     // this check ensures that the two files are consistent ie. if we don't have checksums only the rest needs to match we are just
                     // verifying that we are consistent on both ends source and target
                     final boolean hashAndLengthEqual = (
-                            local.checksum() == null
-                                    && remote.checksum() == null
+                            local.checksum().equals(StoreFileMetaData.UNKNOWN_CHECKSUM)
```

Comment:
why do we need this here? I think this entire `hashAndLengthEqual` can and has to go away